pandora-server-directory/src/spaces/space.ts

```diff
@@ -371,6 +408,13 @@ export class Space {
 		if (this.isBanned(character.baseInfo.account))
 			return 'noAccess';
 
+		// If you are on the allow list, you can enter the space
+		if (this.config.allow.includes(character.baseInfo.account.id))
+			return 'ok';
+
+		if (!this.isInUse())
+			return 'spaceNotInUse';
```

Comment:
no this is not specific to invites, though it is invites that give UI access to this option

currently you can enter a space that has no loaded character inside even if you are not an admin, this aims to universally prevent it, before invites you could do it by manually crafting the join with the inactive space id

Reply:
Okay, I walked a bit more through the logic changes this change brings. There are a few cases where you can get to this line and if it wasn't there still be allowed to join (possibly):
- If the room is password protected and you know the password. This is removed by this change. I would argue that we should allow people that know the password to enter even if the room is not loaded, as for me password is an alternative mechanism to specifying allowlist - instead of telling Pandora who can join, you define who can join by giving them the password. So I think password and allow should be treated similarly.
- If the room is public, but has no admin inside. I honestly think this might be an existing bug and the check should use `this.isPublic` instead of `config.public`. `hasAdminInside` Also should be updated to use `c.isOnline()` instead of `c.assignedClient != null` to be more consistent.
- If the person has a valid invite, they should be able to join with it. (this change prevents that iff there is noone online in the room) On the other hand I think it would be better to have invite check that the person who created the invite is currently in the room for it to still be valid. To me that feels like a more reliable behavior than "anyone is online".

Reply:
Claudia changed my mind about the second point. Public room without anyone inside should be still freely joinable so long you know the ID. It should only be unlisted on the public search.
Either way this change would prevent this.

Reply:
I agree with the first one. But as the space would not be listed, you would need to know the ID and use the general link or an invite link if you have one to get to the interface where you can enter the password to join.

About the 3rd point: I also think that knowing the ID of a space and using it in a general join URL can bring up the join interface without showing the extended information and if the space is public but unlisted, you should not be able to join, does not matter if no one is online, or someone is online but no admin inside. This is to protect a public space without an admin from people who may spam the chat, or disturb in other ways, e.g. by moving their character around wildly.

If you have a special invite that allows you specifically or anyone who has the invite to join the space, then I think it would be confusing though, if you can only use it to join if the one who made the invite is currently inside the space (online or not). I think if you have a special valid invite link (not the one anyone can make by copying the space ID into an URL) you should always be able to join, even if the room unlisted (empty or not), unless of course the room is password protected and your invite link does not allow to bypass that.

One thought I had in addition was that it could be nice if any person inside a public room that is not password protected could still invite someone they know through DMs to an unlisted room they are playing inside in currently. But this would not be a priority.

Reply:
The last two parts can be summarized as: You can join if you have a valid invite link.

Next would be a discussion about what is a "valid" invite link. But I would leave that for another PR, as this one focuses on adding allowlist.

Personally I would say that invite can specify if it is valid even without the one who created it is inside the space. Only admins+ would be able to create invite that is valid without them in the space, but anyone could create an invite that is valid only while they are still inside the space.
To me that makes sense as you want to be able to invite friend somewhere where you already are (of course would only work if you can join the room yourself normally, e.g. the room must be public (even if unlisted) - if it has strict whitelist-only entry, then you can't invite someone who isn't whitelisted unless you are an admin and so can add them in the first place), but we want to avoid person being able to create invite that can be used even long after they leave, which would be unpleasant surprise for those left inside the room who assume it is a private room.
Doing this might however need adjustments to how invites are limited as having constant limit of n invites per space would make it easy to DoS being able to create invites in this case.

Reply:
Yes, I fully agree with this last comment!

And the generic https://project-pandora.com/space/join/safgwsuzagasjasdfsafsa is from my point of view not a valid invite link to join unless the room is listed fully public with no restrictions and anyone can join anyway.

Reply:
for this PR I limit changes to this

- public rooms can only be joined if they are publicly listed
- you can always join with password
- unique invite will be able to show extended info all the time, but password protected room need bypassPassword flag in the invite

for the first point, I am unsure if shard type stable should be checked or not, I feel natural about it, for now I inclined to not check this

Reply:
also not part of this PR but UI for joining unlisted password protected rooms is missing, so the only way to do so is manual for now

Reply:
> also not part of this PR but UI for joining unlisted password protected rooms is missing, so the only way to do so is manual for now

I do not think there will be any UI for joining unlisted password protected spaces.
You either know the link to it or not and that is fine.
But we could open a new issue about a feature idea that Jomshir brought up yesterday: Letting users pin their favorite public spaces to some favorite spaces screen, which would help with this case sufficiently in my view.